File: android/app/src/main/java/com/example/pandu_navigation/data/NavigationDao.java
package com.example.pandu_navigation.data;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.OnConflictStrategy;
import java.util.List;

@Dao
public interface NavigationDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertBreadcrumb(BreadcrumbEntity breadcrumb);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertTrails(List<TrailEntity> trails);

    // Bounding Box Query: Find trails that physically overlap with a buffer around
    // the user
    // The trail's bounds must intersect the user's buffer.
    // Simplification: Select trails where (trail.minLat <= userLat+buffer) AND
    // (trail.maxLat >= userLat-buffer) ...
    // Actually, prompt asked for "minLat >= :lat - :buffer" etc.
    // The user wants trails STRICTLY RELEVANT.
    // A trail is relevant if its bounding box OVERLAPS the search window.

    @Query("SELECT * FROM trails WHERE " +
            "max_lat >= (:lat - :buffer) AND " +
            "min_lat <= (:lat + :buffer) AND " +
            "max_lng >= (:lng - :buffer) AND " +
            "min_lng <= (:lng + :buffer)")
    List<TrailEntity> getNearbyTrails(double lat, double lng, double buffer);

    @Query("SELECT * FROM trails")
    List<TrailEntity> getAllTrails();

    @Query("DELETE FROM trails")
    void clearTrails();

    @Query("DELETE FROM user_breadcrumbs")
    void clearBreadcrumbs();
}

File: PanduNavigationNative/app/src/main/java/com/example/gnav/data/db/NavigationDao.kt
package com.example.gnav.data.db

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

@Dao
interface NavigationDao {
    // Seeding
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertMountains(mountains: List<MountainEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertTrails(trails: List<TrailEntity>)

    @Query("SELECT COUNT(*) FROM mountains")
    suspend fun getMountainCount(): Int

    // Queries
    @Query("SELECT * FROM mountains")
    fun getMountains(): Flow<List<MountainEntity>>

    @Query("SELECT * FROM trails WHERE mountainId = :mountainId")
    suspend fun getTrailsByMountain(mountainId: String): List<TrailEntity>

    @Query("SELECT * FROM trails WHERE id = :trailId LIMIT 1")
    suspend fun getTrail(trailId: String): TrailEntity?

    // Tracking
    @Insert
    suspend fun insertBreadcrumb(breadcrumb: BreadcrumbEntity)

    @Query("SELECT * FROM breadcrumbs ORDER BY timestamp DESC LIMIT 1")
    fun getLastBreadcrumb(): Flow<BreadcrumbEntity?>
}
